rfcs/0049-flakes.md

```diff
@@ -0,0 +1,703 @@
+
+A number of outputs have a specific meaning to Nix or other tools like
+Hydra. Currently, these are:
+
+* `packages`: A set of derivations used as a default by most `nix`
+  commands. For example, `nix run nixpkgs:hello` uses the
+  `packages.hello` attribute of the `nixpkgs` flake. It cannot contain
+  any non-derivation attributes. This also means it cannot be a nested
+  set! (The rationale is that supporting nested sets requires Nix to
+  evaluate each attribute in the set, just to discover which packages
+  are provided.)
+
+* `defaultPackage`: A derivation used as a default by most `nix`
+  commands if no attribute is specified. For example, `nix run
+  dwarffs` uses the `defaultPackage` attribute of the `dwarffs` flake.
+
+* `checks`: A non-nested set of derivations built by the `nix flake
+  check` command, and by Hydra if a flake does not have a `hydraJobs`
+  attribute.
+
+* `hydraJobs`: A nested set of derivations built by Hydra.
+
+* `devShell`: A derivation that defines the shell environment used by
+  `nix dev-shell` if no specific attribute is given. If it does not
+  exist, then `nix dev-shell` will use `defaultPackage`.
+
+TODO: NixOS-related outputs such as `nixosModules` and `nixosSystems`.
+
+## Flake references
+
+Flake references are a vaguely URL-like syntax to specify which flake
+to use. This is used on the command line (e.g. in `nix build
+nixpkgs:hello`, `nixpkgs` is a flake reference), and in the list of
+flake dependencies in `flake.nix` (e.g. in `inputs = [ "nixpkgs" ];`).
+
+Currently the following types of flake references are supported:
+
+* Git repositories. These have the form
+
+      (http|https|ssh|git|file):(//<server>)?<path>(\?<params>)?
+
+  with the constraint that `<path>` must end with `.git` for
+  non-`file` repositories. `<params>` are a list of key/value pairs in
+  URI query parameter syntax. The following parameters are supported:
+
+  * `ref`: The branch or tag to fetch. The default is `master`.
+  * `rev`: The Git commit hash to fetch. Note that this commit must be
+    an ancestor of `ref`, since Nix doesn't clone the entire
+    repository, only the specified `ref` (and the Git protocol doesn't
+    allow fetching a `rev` without a known `ref`). The default is the
+    commit currently pointed to by `ref`.
+  * `dir`: The subdirectory of the repository in which `flake.nix` is
+    located. This parameter enables having multiple flakes in a
+    repository. The default is the root directory.
+
+  For example, the following are valid Git flake references:
+
+  * `https://example.org/my/repo.git`
+  * `https://example.org/my/repo.git?dir=flake1`
+  * `ssh://git@github.com:NixOS/nix.git?ref=v1.2.3`
+  * `git://github.com/edolstra/dwarffs.git?ref=unstable&rev=e486d8d40e626a20e06d792db8cc5ac5aba9a5b4`
+  * `file:///home/my-user/some-repo/some-repo.git`
+
+* Local paths. These have the form
+
+      <path>(\?<params)?
+
+  where `<path>` must refer to (a subdirectory of) a Git repository.
```

Comment:
Can we please support plain directories? Then we don't need to wait until at least Mercurial is supported, not to mention negotiating about Fossil, Pijul, Monotone…

Reply:
I think this at least is worth thinking about before we make any sort of flakes release, because I think otherwise it could cause compatibility problems down the line, especially in regards to:


>    * The `dir` parameter is automatically derived. For example, if `/foo/bar` is a Git repository, then the flake reference `/foo/bar/flake` is equivalent to `/foo/bar?dir=flake`.

How would we resolve this? Would such a directory be a git flake with a `dir` parameter, or a `dir` repository. What if you wanted the other? Would there be a way to override it?

Reply:
I think it’s also important that it would at least be possible to implement fetching remote tarballs, even if we don’t support it right away. I don’t see any immediate compatibility concerns with that, though.

Reply:
requiring local paths to be git repos would be annoying (how would this behave when nesting git repositories?) and there's nothing stopping me from having a `?` inside a path; not being able to tab-complete those would be a pain..

Reply:
not only that, but my current workflow (nixpkgs with custom patches inside a git subtree) is heavily discouraged, as this means i have a `flakes.nix` not in the root of a git repo, which the spec only barely seems to support.